source/shared/cpp/ObjectModel/BaseElement.h

```diff
@@ -0,0 +1,127 @@
+#pragma once
+
+#include "pch.h"
+#include "json/json.h"
+#include "ParseContext.h"
+#include "ParseUtil.h"
+#include "SemanticVersion.h"
+#include "RemoteResourceInformation.h"
+
+namespace AdaptiveSharedNamespace
+{
+    class BaseElement
+    {
+    public:
+        BaseElement() :
+            m_additionalProperties{}, m_typeString{}, m_requires(0), m_fallbackContent(nullptr),
+            m_fallbackType(FallbackType::None), m_id{}, m_internalId{GetNextInternalId()}
+        {
+            PopulateKnownPropertiesSet();
+        }
+
+        BaseElement(const BaseElement&) = default;
+        BaseElement(BaseElement&&) = default;
+        BaseElement& operator=(const BaseElement&) = default;
+        BaseElement& operator=(BaseElement&&) = default;
+        virtual ~BaseElement() = default;
+
+        // Element type and identity
+        std::string GetElementTypeString() const;
+        void SetElementTypeString(const std::string& value);
+        virtual std::string GetId() const;
+        virtual void SetId(const std::string& value);
+        unsigned int GetInternalId() const { return m_internalId; }
+
+        // Element parsing/serialization
+        template<typename T>
+        static void ParseJsonObject(AdaptiveSharedNamespace::ParseContext& context,
+                                    const Json::Value& json,
+                                    std::shared_ptr<BaseElement>& baseElement)
+        {
+            T::ParseJsonObject(context, json, baseElement);
+        }
+
+        template<typename T> void DeserializeBase(ParseContext& context, const Json::Value& json);
+
+        virtual std::string Serialize() const;
+        virtual Json::Value SerializeToJsonValue() const;
+        Json::Value GetAdditionalProperties() const;
+        void SetAdditionalProperties(const Json::Value& additionalProperties);
+
+        FallbackType GetFallbackType() const { return m_fallbackType; }
+        std::shared_ptr<BaseElement> GetFallbackContent() const { return m_fallbackContent; }
+
+        bool MeetsRequirements(const std::unordered_map<std::string, std::string>& hostProvides) const;
+
+        // misc
+        virtual void GetResourceInformation(std::vector<RemoteResourceInformation>& resourceUris);
+
+    protected:
+        // Element parsing/serialization
+        virtual void PopulateKnownPropertiesSet();
+        void SetTypeString(const std::string& type) { m_typeString = type; }
+        std::string m_typeString;
+        std::unordered_set<std::string> m_knownProperties;
+        Json::Value m_additionalProperties;
+
+    private:
+        static unsigned int GetNextInternalId()
+        {
+            static unsigned int nextId = 1;
+            return nextId++;
```

Comment:
shouldn't we think of overflow strategy since renderers on mobile app usually sleeps and rarely quit?

Reply:
I'm not too concerned about the overflow collision scenario -- you'd have to have more than 4,294,967,295 elements in a single card to run into a collision. It seems exceedingly unlikely that we'd manage to construct that many elements in a single session. However, there *is* one small bug here that I need to fix -- if we overflow back to 0, we'll hand out our `InvalidId` sentinel value.